File: android/src/main/kotlin/com/michaellazebny/jyphoon/jc/methods/MediaChannel.kt
package com.michaellazebny.jyphoon.jc.methods

import com.juphoon.cloud.JCMediaChannel
import com.juphoon.cloud.JCMediaDevice
import com.michaellazebny.jyphoon.jc.JCWrapper.JCManager
import com.michaellazebny.jyphoon.jc.tools.JCCallUtils

class MediaChannel {
    private val jcManager = JCManager.getInstance()

    fun join(channelId: String, password: String, video: Boolean): Boolean {
        // 生成 join 参数
        val joinParam = JCMediaChannel.JoinParam()
        joinParam.capacity = 2
        jcManager.mediaDevice.enableSpeaker(true)
        jcManager.mediaChannel.enableUploadAudioStream(true)
        jcManager.mediaChannel.enableUploadVideoStream(true)
        val res = jcManager.mediaChannel.join(channelId, joinParam)
        if (res && video) {
            setVideo(true)
        }
        return res
    }

    fun leave(): Boolean {
        return jcManager.mediaChannel.leave()
    }

    fun audio(): Boolean {
        val selfAudio = jcManager.mediaChannel.selfParticipant?.isAudio ?: false
        val audioStart = jcManager.mediaDevice.isAudioStart
        val uploadLocalAudio = jcManager.mediaChannel.uploadLocalAudio
        return selfAudio && audioStart && uploadLocalAudio
    }

    fun video(): Boolean {
        val selfVideo = jcManager.mediaChannel.selfParticipant?.isVideo ?: false
        val videoStart = jcManager.mediaDevice.isCameraOpen
        val uploadLocalVideo = jcManager.mediaChannel.uploadLocalVideo
        return selfVideo && videoStart && uploadLocalVideo
    }

    fun otherAudio(): Boolean {
        return JCCallUtils.otherParticipant?.isAudio ?: false
    }

    fun otherVideo(): Boolean {
        return JCCallUtils.otherParticipant?.isVideo ?: false
    }

    fun setAudio(audio: Boolean) {
        jcManager.mediaChannel.enableUploadAudioStream(audio)
        if (audio) {
            jcManager.mediaDevice.startAudio()
        } else {
            jcManager.mediaDevice.stopAudio()
        }
    }

    fun setVideo(video: Boolean) {
        jcManager.mediaChannel.enableUploadVideoStream(video)
        if (video) {
            jcManager.mediaDevice.startCamera()
        } else {
            jcManager.mediaDevice.stopCamera()
        }
    }

    fun setSpeaker(value: Boolean) {
        jcManager.mediaDevice.enableSpeaker(value)
    }

    fun confStatus(): String {
        val participants = jcManager?.mediaChannel?.participants
        if (participants == null || participants.isEmpty()) {
            return "off"
        }
        if (participants.size == 1) {
            return "waiting"
        }
        return "on"
    }

    fun switchCamera() {
        jcManager.mediaDevice.switchCamera()
    }
}

File: ios/Classes/Methods/MediaChannel.swift
import Foundation
import JCSDKOC.JCMediaChannel

class MediaChannel {
    private let jc = JCRoom.shared
    /// Must be called after initialization
    public func join(channelId: String, password: String, video: Bool, asr: Bool) -> Bool {
        let joinParam = JCMediaChannelJoinParam()
        joinParam.capacity = 2
        jc.mediaDevice.enableSpeaker(true)
        jc.mediaChannel.enableUploadAudioStream(true)
        jc.mediaChannel.enableUploadVideoStream(true)
        let userID = channelId.components(separatedBy: "_").last
        let jcCallParam = JCCallParam(extraParam: video ? "video" : "audio", ticket: channelId)
        let res = asr ?
        jc.call.call(userID!, video: video, callParam: jcCallParam) :
        jc.mediaChannel.join(channelId, joinParam: joinParam)
        if (res && video) {
            let videoRes = setVideo(video: true)
            print("VIDEORES: \(videoRes)")
        }
        return res
    }
    
    public func leave() -> Bool {
        return JCRoom.shared.mediaChannel.leave()
    }
    
    public func audio() -> Bool {
        let selfAudio = jc.mediaChannel.selfParticipant?.audio ?? false
        let audioStart = jc.mediaDevice.audioStart
        let uploadLocalAudio = jc.mediaChannel.uploadLocalAudio
        return selfAudio && audioStart && uploadLocalAudio
    }
    
    public func video() -> Bool {
        let selfVideo = jc.mediaChannel.selfParticipant?.video ?? false
        let cameraOpen = jc.mediaDevice.cameraOpen
        let uploadLocalVideo = jc.mediaChannel.uploadLocalVideo
        return selfVideo && cameraOpen && uploadLocalVideo
    }
    
    public func otherAudio() -> Bool {
        JCRoomUtils.otherParticipant?.audio ?? false
    }
    
    public func otherVideo() -> Bool {
        JCRoomUtils.otherParticipant?.video ?? false
    }
    
    public func setAudio(audio: Bool) {
        JCRoom.shared.mediaChannel.enableUploadAudioStream(audio)
    }
    
    public func setVideo(video: Bool) -> Bool {
        JCRoom.shared.mediaChannel.enableUploadVideoStream(video)
        if (video) {
            return JCRoom.shared.mediaDevice.startCamera()
        } else {
            return JCRoom.shared.mediaDevice.stopCamera()
        }
    }
    
    public func setSpeaker(speaker: Bool) {
        JCRoom.shared.mediaDevice.enableSpeaker(speaker)
    }
    
    public func confStatus() -> String {
        let participants = JCRoom.shared.mediaChannel.participants
        if (participants.isEmpty) {
            return "off"
        }
        if (participants.count == 1) {
            return "waiting"
        }
        return "on"
    }
    
    public func switchCamera() {
        JCRoom.shared.mediaDevice.switchCamera()
    }
}
